### PR TITLE
fix: harden ElGamal decryption limits and execution timing

### PR DESCRIPTION
**Pull request by mrtcnk**
Key Changes:

ElGamal Decryption & Side-Channels
- Eliminated the zero-amount fast path in the decryption loop
- Rewrote secp256k1_solve_dlp_small_range_fixed to use a fixed iteration
  count (exactly 1,000,000) with branchless bitwise operations; documented
  reliance on variable-time libsecp256k1 internals
- Added boundary condition unit tests (1,000,000 vs 1,000,001)
- Updated API docs to reflect the 1M recovery limit; recommended BSGS for
  larger balances in off-chain computations

Deprecated OpenSSL API & Hardening
- Migrated all deprecated hashing functions (SHA256_Init, etc.) to the
  modern FIPS-compliant EVP_MD_CTX interface
- Enforced strict return value checks (!= 1) on all EVP operations
- Hardened challenge generation across all Sigma protocols to fail
  deterministically and poison output buffers (memset) on error
- Removed redundant context resets

Domain Tags & Architecture
- Removed standalone plaintext equality proofs (proof_same_plaintext.c
  and proof_same_plaintext_multi.c)
- Variant A linkage will be moved into the shared-randomness architecture;
  standalone proof dedicated exclusively to Variant B

Bulletproofs Side-Channels
- Replaced variable-time memcmp calls with a constant-time byte-OR
  helper (scalar_is_zero)
- Removed cache-timing vulnerabilities in bit decomposition logic using
  branchless two's complement bitmasks

Aggregated Bulletproof MSM Crash
- Hardened MSM logic to safely handle the point at infinity
- Introduced accumulator flag (acc_inited) to prevent illegal operations
  on edge-case vectors (e.g., all-zero or all-MAX amounts)

Participant Bounds & Memory Safety
- Enforced strict internal limits (n <= 4 for equality, m <= 4 for
  aggregation) to prevent integer overflows and unbounded allocations

Serialization Integrity
- Wrapped all secp256k1_ec_pubkey_serialize and commitment creation calls
  (including inside SER_AND_HASH) with strict length assertions (len == 33)
  and return value checks before hashing

### CODE REVIEW

**Comment by mathbunnyru** [comment]
@tesseract-ripple @akinyele-ripple please review

**Review by tesseract-ripple — COMMENTED** [review]
test

**Review by tesseract-ripple — COMMENTED** [review]
Looks good overall. Verified Fiat-Shamir transcript orderings (prover vs verifier consistent for y, z, x, IPA), traced the branchless DLP logic through all cases, and checked the new bit decomposition, serialize checks, and participant bounds. Inline comments on specific lines below.

Two file-level notes:

- **`src/proof_link.c`**: uses `EVP_MD_CTX` / `EVP_DigestInit_ex` but doesn't include `<openssl/evp.h>` directly — works via transitive includes from `<openssl/sha.h>`, but that's fragile. Add the include and consider dropping `sha.h`.

- **Challenge hash functions** (`compute_challenge_equality` in `equality_proof.c`, `build_link_challenge_hash`, `build_pok_challenge`, `compute_challenge_equality_shared_r`) return `void`, so EVP failures leave `e_out` unwritten and callers use whatever was on the stack. Low practical risk since EVP basically never fails, but worth a follow-up to return `int` so callers can propagate errors.

The API deletion of the standalone plaintext equality proofs is fine — we confirmed rippled production code uses only the `_shared_r` variant. One test helper (`test/jtx/impl/mpt.cpp:1386`) references `secp256k1_mpt_prove_same_plaintext_multi_size` for a dummy buffer and will need a trivial rippled-side fix.

File: include/secp256k1_mpt.h [hunk 1]
```diff
@@ -31,6 +31,16 @@ secp256k1_elgamal_encrypt(
 
 /**
  * @brief Decrypts an ElGamal ciphertext to recover the amount.
+ * Due to the use of a linear discrete logarithm search, this function can
+ * only successfully recover plaintext amounts between 0 and 1,000,000 (inclusive).
```
Comment (tesseract-ripple) [inline include/secp256k1_mpt.h:35]: The commit message describes this as "strict constant time" but the underlying libsecp256k1 operations are variable-time: `ec_pubkey_combine` calls `gej_add_ge_var` internally, `ec_pubkey_serialize` timing depends on field normalization, and there are data-dependent branches in the loop (`if (combine_ok)`). The fixed-iteration-count loop is a meaningful improvement over the early-exit version, but suggest updating the commit message and this doc comment to say "fixed iteration count" rather than "strict constant time" — we don't want downstream consumers relying on a CT guarantee the library layer can't deliver. (wallets etc. that need to decrypt large amounts should use a more efficient DLP solver anyway).

File: src/bulletproof_aggregated.c [hunk 2]
```diff
@@ -203,28 +219,32 @@ int secp256k1_bulletproof_ipa_msm(const secp256k1_context *ctx,
...
   int initialized = 0;
-  unsigned char zero[32] = {0};
 
   for (size_t i = 0; i < n; ++i)
   {
-    if (memcmp(scalars + i * 32, zero, 32) == 0)
+    unsigned char s_tmp[32];
+    memcpy(s_tmp, scalars + i * 32, 32);
+
+    if (scalar_is_zero(s_tmp))
```
Comment (tesseract-ripple) [inline src/bulletproof_aggregated.c:232]: The `scalar_is_zero` check is now constant-time (byte-OR accumulation), but the `continue` branch on the result is still data-dependent. In the IPA prover (round 0), these scalars are the bit decomposition of the secret value. The comment above documents this as accepted risk, which seems right for the threat model, just flagging for the record.

File: src/bulletproof_aggregated.c [hunk 3]
```diff
@@ -44,6 +44,7 @@
  */
 #include "secp256k1_mpt.h"
 #include <assert.h>
+#include <openssl/evp.h>
 #include <openssl/rand.h>
 #include <openssl/sha.h>
```
Comment (tesseract-ripple) [inline src/bulletproof_aggregated.c]: Dead include — `<openssl/sha.h>` is unused after the EVP migration.

File: src/commitments.c [hunk 4]
```diff
@@ -61,39 +62,49 @@ int secp256k1_mpt_hash_to_point_nums(const secp256k1_context *ctx,
...
-    SHA256_Init(&sha);
-    SHA256_Update(&sha, "MPT_BULLETPROOF_V1_NUMS", 23); // Domain Sep
-    SHA256_Update(&sha, "secp256k1", 9);                // Curve Label
+    EVP_MD_CTX_reset(mdctx);
+    if (EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL) != 1)
+    {
+      EVP_MD_CTX_free(mdctx);
+      return 0;
+    }
+    EVP_DigestUpdate(mdctx, "MPT_BULLETPROOF_V1_NUMS", 23); // Domain Sep
```
Comment (tesseract-ripple) [inline src/commitments.c]: These two `EVP_DigestUpdate` calls don't check return values, unlike every other EVP call in the PR.

File: src/equality_proof.c [hunk 5]
```diff
@@ -90,50 +90,82 @@ static void compute_challenge_equality(
...
   unsigned char h[32];
   size_t len;
   const char *domain = "MPT_POK_PLAINTEXT_PROOF";
 
-  SHA256_Init(&sha);
-  SHA256_Update(&sha, domain, strlen(domain));
+  if (!mdctx)
+    return;
+
+  EVP_MD_CTX_reset(mdctx);
```
Comment (tesseract-ripple) [inline src/equality_proof.c]: `EVP_MD_CTX_reset` immediately after `EVP_MD_CTX_new` is a no-op — `new()` already returns a clean context. Same pattern in `proof_link.c`, `proof_pok_sk.c`, and `proof_same_plaintext_multi_shared_r.c`.

File: src/bulletproof_aggregated.c [hunk 6]
```diff
@@ -723,66 +724,94 @@ int derive_ipa_binding_challenge(const secp256k1_context *ctx,
...
+cleanup:
+  EVP_MD_CTX_free(mdctx);
+  return ok;
 }
+
 /**
  * Derive u = H(last_challenge || L || R) reduced to a valid scalar.
  * IMPORTANT: use the SAME exact logic in verifier.
  */
+// C
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/bulletproof_aggregated.c]: Done!

File: src/proof_same_plaintext_multi_shared_r.c [hunk 7]
```diff
@@ -90,56 +90,72 @@ static void compute_challenge_equality_shared_r(
...
                                   SECP256K1_EC_COMPRESSED);
-    SHA256_Update(&sha, buf, 33);
+    if (EVP_DigestUpdate(mdctx, buf, 33) != 1)
+      goto cleanup;
     len = 33;
     secp256k1_ec_pubkey_serialize(ctx, buf, &len, &Pk_vec[i],
                                   SECP256K1_EC_COMPRESSED);
-    SHA256_Update(&sha, buf, 33);
+    if (EVP_DigestUpdate(mdctx, buf, 33) != 1)
+      goto cleanup;
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_same_plaintext_multi_shared_r.c:138]: Done!

File: src/proof_pok_sk.c [hunk 8]
```diff
@@ -70,30 +70,43 @@ static void build_pok_challenge(const secp256k1_context *ctx,
...
   }
 
-  SHA256_Final(h, &sha);
+  if (EVP_DigestFinal_ex(mdctx, h, NULL) != 1)
+    goto cleanup;
   secp256k1_mpt_scalar_reduce32(e_out, h);
+
+cleanup:
+  EVP_MD_CTX_free(mdctx);
 }
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_pok_sk.c:126]: I have updated the internal helper.

File: src/proof_same_plaintext_multi.c [hunk 9]
```diff
@@ -76,68 +76,86 @@ static int generate_random_scalar(const secp256k1_context *ctx,
 /*
  * Hash( Domain || {R_i, S_i, Pk_i} || Tm || {TrG_i, TrP_i} || TxID )
  */
+// C
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_same_plaintext_multi.c]: Done!

File: src/equality_proof.c [hunk 10]
```diff
@@ -90,50 +90,64 @@ static void compute_challenge_equality(
...
   }
 
-  SHA256_Final(h, &sha);
+  if (EVP_DigestFinal_ex(mdctx, h, NULL) != 1)
+    goto cleanup;
   secp256k1_mpt_scalar_reduce32(e_out, h);
+
+cleanup:
+  EVP_MD_CTX_free(mdctx);
 }
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/equality_proof.c:180]: I've updated.

File: src/proof_link.c [hunk 11]
```diff
@@ -80,23 +80,29 @@ static void build_link_challenge_hash(
...
-  SHA256_Init(&sha);
-  SHA256_Update(&sha, domain, strlen(domain));
+  if (!mdctx)
+    return;
+
+  EVP_MD_CTX_reset(mdctx);
+  if (EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL) != 1)
+    goto cleanup;
+  if (EVP_DigestUpdate(mdctx, domain, strlen(domain)) != 1)
+    goto cleanup;
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_link.c:100]: Done!

File: src/proof_same_plaintext_multi_shared_r.c [hunk 12]
```diff
@@ -90,56 +90,72 @@ static void compute_challenge_equality_shared_r(
...
   }
 
-  SHA256_Final(h, &sha);
+  if (EVP_DigestFinal_ex(mdctx, h, NULL) != 1)
+    goto cleanup;
   secp256k1_mpt_scalar_reduce32(e_out, h);
+
+cleanup:
+  EVP_MD_CTX_free(mdctx);
 }
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_same_plaintext_multi_shared_r.c:181]: I have updated.

File: src/commitments.c [hunk 13]
```diff
@@ -48,6 +48,7 @@
  * @param index     Vector index (enforced Big-Endian).
  * @return 1 on success, 0 on failure.
  */
+// C
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/commitments.c]: Done!

File: src/equality_proof.c [hunk 14]
```diff
@@ -90,50 +90,64 @@ static void compute_challenge_equality(
...
   len = 33;
   secp256k1_ec_pubkey_serialize(ctx, buf, &len, c2, SECP256K1_EC_COMPRESSED);
-  SHA256_Update(&sha, buf, 33);
+  if (EVP_DigestUpdate(mdctx, buf, 33) != 1)
+    goto cleanup;
   len = 33;
   secp256k1_ec_pubkey_serialize(ctx, buf, &len, pk, SECP256K1_EC_COMPRESSED);
-  SHA256_Update(&sha, buf, 33);
+  if (EVP_DigestUpdate(mdctx, buf, 33) != 1)
+    goto cleanup;
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/equality_proof.c:133]: I have addressed this.

File: src/elgamal.c [hunk 15]
```diff
@@ -122,65 +122,177 @@ int secp256k1_elgamal_encrypt(const secp256k1_context *ctx,
...
-    if (!secp256k1_ec_pubkey_combine(ctx, &next_M, pts, 2))
-      return 0;
-    current_M = next_M;
+  /* 4. Call the Modular DLP Solver */
+  uint64_t loop_amount = 0;
+  uint64_t match_loop = 0;
+
+  if (!secp256k1_solve_dlp_small_range_ct(ctx, &loop_amount, &match_loop,
+                                          M_target_ser, 1000000))
+  {
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/elgamal.c:305]: I added test_decryption_boundaries to explicitly encrypt and attempt decryption on both 1,000,000 and 1,000,001.

File: src/bulletproof_aggregated.c [hunk 16]
```diff
@@ -1339,77 +1368,164 @@ int secp256k1_bulletproof_prove_agg(const secp256k1_context *ctx,
     goto cleanup;
 
   /* ---- 6. Fiat–Shamir y,z ---- */
+  // src/bulletproof_aggregated.c
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/bulletproof_aggregated.c]: Done!

[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/bulletproof_aggregated.c]: Done!

File: include/secp256k1_mpt.h (hunk 1 above)
Comment (mrtcnk) [reply include/secp256k1_mpt.h:35]: Thanks for the comment! I've updated the header documentation and renamed the internal function to ..._fixed to clearly specify that this is a 'fixed iteration count' search.

File: src/bulletproof_aggregated.c (hunk 2 above)
Comment (mrtcnk) [reply src/bulletproof_aggregated.c:232]: Thank you for the note!

File: src/commitments.c (hunk 4 above)
Comment (mrtcnk) [reply src/commitments.c]: Great catch, thanks! Done!

File: src/bulletproof_aggregated.c (hunk 3 above)
Comment (mrtcnk) [reply src/bulletproof_aggregated.c]: Done!

File: src/equality_proof.c (hunk 5 above)
Comment (mrtcnk) [reply src/equality_proof.c]: Thanks for the comment. Done!

File: src/elgamal.c [hunk 17]
```diff
@@ -122,65 +122,177 @@ int secp256k1_elgamal_encrypt(const secp256k1_context *ctx,
...
+    pts[0] = &current_M;
+    pts[1] = &G_point;
+    int combine_ok = secp256k1_ec_pubkey_combine(ctx, &next_M, pts, 2);
+
+    /* On failure, freeze current_M in place. Loop must always run to max_range.
+     */
+    if (combine_ok)
+    {
+      current_M = next_M;
+    }
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/elgamal.c:213]: I have updated.

File: src/proof_pok_sk.c [hunk 18]
```diff
@@ -70,30 +70,43 @@ static void build_pok_challenge(const secp256k1_context *ctx,
...
   secp256k1_ec_pubkey_serialize(ctx, buf, &len, pk, SECP256K1_EC_COMPRESSED);
-  SHA256_Update(&sha, buf, 33);
+  if (EVP_DigestUpdate(mdctx, buf, 33) != 1)
+    goto cleanup;
 
   len = 33;
   secp256k1_ec_pubkey_serialize(ctx, buf, &len, T, SECP256K1_EC_COMPRESSED);
-  SHA256_Update(&sha, buf, 33);
+  if (EVP_DigestUpdate(mdctx, buf, 33) != 1)
+    goto cleanup;
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_pok_sk.c:105]: I have updated.

File: src/equality_proof.c [hunk 19]
```diff
@@ -42,8 +42,8 @@
  */
 
 #include "secp256k1_mpt.h"
+#include <openssl/evp.h>
 #include <openssl/rand.h>
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/equality_proof.c:47]: Done!

File: src/elgamal.c [hunk 20]
```diff
@@ -36,8 +36,8 @@
  * @see [Spec (ConfidentialMPT_20260201.pdf) Section 3.2.2] ElGamal Encryption
  */
 #include "secp256k1_mpt.h"
+#include <openssl/evp.h>
 #include <openssl/rand.h>
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/elgamal.c:41]: Done!

File: src/proof_pok_sk.c [hunk 21]
```diff
@@ -39,8 +39,8 @@
  * of Secret Key
  */
 #include "secp256k1_mpt.h"
+#include <openssl/evp.h>
 #include <openssl/rand.h>
-#include <openssl/sha.h>
 #include <stdlib.h>
 #include <string.h>
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_pok_sk.c:47]: Done!

File: src/elgamal.c [hunk 22]
```diff
@@ -122,65 +122,184 @@ int secp256k1_elgamal_encrypt(const secp256k1_context *ctx,
...
+    /* Expand to 64-bit before the idiom to make width explicit.
+     * Nonzero detection: if diff64 != 0, saturate bit 63. */
+    uint64_t diff64 = (uint64_t)match_diff;
+    uint64_t match = 1 ^ (((diff64 | (~diff64 + 1)) >> 63) & 1);
+
+    /* Constant-time assignment mask */
+    uint64_t mask = ~(match - 1);
+    found_amount ^= (found_amount ^ i) & mask;
+    is_found |= match;
+
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/elgamal.c:196]: Done!

File: src/elgamal.c [hunk 23]
```diff
@@ -122,65 +122,184 @@ int secp256k1_elgamal_encrypt(const secp256k1_context *ctx,
...
-  /* This is much faster than doing point subtraction first */
-  if (pubkey_equal(ctx, c2, &S))
+  ser_len = 33;
+  if (!secp256k1_ec_pubkey_serialize(ctx, c2_ser, &ser_len, c2,
+                                     SECP256K1_EC_COMPRESSED))
+    return 0;
+  ser_len = 33;
+  if (!secp256k1_ec_pubkey_serialize(ctx, S_ser, &ser_len, &S,
+                                     SECP256K1_EC_COMPRESSED))
+    return 0;
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/elgamal.c:260]: Done!

File: src/proof_same_plaintext_multi_shared_r.c [hunk 24]
```diff
@@ -50,8 +50,8 @@
  * Plaintexts with Shared Randomness
  */
 #include "secp256k1_mpt.h"
+#include <openssl/evp.h>
 #include <openssl/rand.h>
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_same_plaintext_multi_shared_r.c:55]: Done!

File: tests/test_ipa.c [hunk 25]
```diff
@@ -220,17 +221,46 @@ int main(void)
...
+
+    /* ux = SHA256(ipa_transcript_id || dot) */
+    if (EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL) != 1)
+      goto test_cleanup;
+    if (EVP_DigestUpdate(mdctx, ipa_transcript_id, 32) != 1)
+      goto test_cleanup;
+    if (EVP_DigestUpdate(mdctx, dot, 32) != 1)
+      goto test_cleanup;
+    if (EVP_DigestFinal_ex(mdctx, ux, &md_len) != 1)
+      goto test_cleanup;
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply tests/test_ipa.c:251]: Done!

File: src/proof_link.c [hunk 26]
```diff
@@ -48,8 +48,8 @@
  */
 #include "secp256k1_mpt.h"
 #include <assert.h>
+#include <openssl/evp.h>
 #include <openssl/rand.h>
-#include <openssl/sha.h>
 #include <stdio.h>
 #include <string.h>
```
[the remark this answers is not in the document]
Comment (mrtcnk) [reply src/proof_link.c:56]: Done!

File: src/commitments.c [hunk 27]
```diff
@@ -29,7 +29,8 @@
  */
 #include "secp256k1_mpt.h"
 #include <openssl/crypto.h> // For OPENSSL_cleanse
-#include <openssl/sha.h>
+#include <openssl/crypto.h>
```
Comment (tesseract-ripple) [inline src/commitments.c]: **✅ Resolved in 69d810e1.**

~~Nit: duplicate `#include <openssl/crypto.h>` — line 31 already has it.~~

File: src/proof_same_plaintext_multi_shared_r.c [hunk 28]
```diff
@@ -84,74 +85,114 @@ size_t secp256k1_mpt_proof_equality_shared_r_size(size_t n_recipients)
...
     const secp256k1_context *ctx,
     unsigned char *proof_out, // Caller MUST allocate
-                              // secp256k1_mpt_proof_equality_shared_r_size(n)
+    // secp256k1_mpt_proof_equality_shared_r_size(n)
     uint64_t amount, const unsigned char *r_shared, size_t n,
     const secp256k1_pubkey *C1, const secp256k1_pubkey *C2_vec,
     const secp256k1_pubkey *Pk_vec, const unsigned char *context_id)
 {
+  /* Protocol restricts n to max 4 (Sender, Receiver, Issuer, Auditor) */
+  if (n == 0 || n > 4)
```
Comment (tesseract-ripple) [inline src/proof_same_plaintext_multi_shared_r.c:194]: **✅ Resolved in 07916b67.**

~~This `n <= 4` constraint (and `m <= BP_MAX_VALUES` in the Bulletproof prover/verifier) was not documented in the public header `secp256k1_mpt.h`.~~

Docstrings added for `secp256k1_mpt_prove_equality_shared_r`, `secp256k1_mpt_verify_equality_shared_r`, `secp256k1_bulletproof_prove_agg`, and `secp256k1_bulletproof_verify_agg`.

File: src/elgamal.c [hunk 29]
```diff
@@ -122,65 +123,205 @@ int secp256k1_elgamal_encrypt(const secp256k1_context *ctx,
...
+    /* Increment for next loop */
+    pts[0] = &current_M;
+    pts[1] = &G_point;
+    int combine_ok = secp256k1_ec_pubkey_combine(ctx, &next_M, pts, 2);
+
+    /* 2. Branchless Conditional Move for Point Addition
+     * If combine_ok == 1, combine_mask is 0xFF. If combine_ok == 0, it is 0x00.
+     */
+    unsigned char combine_mask = (unsigned char)(0 - combine_ok);
+    unsigned char *curr_ptr = (unsigned char *)&current_M;
```
Comment (tesseract-ripple) [inline src/elgamal.c:206]: This branchless conditional move casts the opaque `secp256k1_pubkey` to `unsigned char *` for byte-level manipulation. It works today because the struct is `unsigned char data[64]`, but it couples to an internal representation detail.

Since `combine_ok` depends on the public iterator (whether `i*G` is the point at infinity — never true for `i <= 1M`), this branch doesn't leak secret information. A plain `if (combine_ok) current_M = next_M;` would be simpler and not fragile against future libsecp256k1 changes.

(Not a blocker — maintainability note.)

File: src/proof_same_plaintext_multi_shared_r.c [hunk 30]
```diff
@@ -327,9 +375,14 @@ int secp256k1_mpt_verify_equality_shared_r(
   if (!secp256k1_ec_seckey_verify(ctx, s_r))
     goto cleanup;
 
+  /* Fail fast! Strict length check before heavy crypto math */
```
Comment (tesseract-ripple) [inline src/proof_same_plaintext_multi_shared_r.c:378]: Nit: at this point all deserialization has completed (points parsed, scalars verified). The check is a post-parse consistency assertion, not an early-out before work.

```suggestion
  /* Post-parse sanity: verify we consumed exactly the expected proof length */
```

**Review by tesseract-ripple — APPROVED** [review]
Re-reviewed after the 6 new commits (81c1916d through 9e72ee52). All three previous blockers are resolved:

1. **CT claim qualified** — header now explicitly disclaims constant-time guarantees for the underlying curve operations. Accurate.
2. **Challenge functions return int** — all callers check the return value; `e_out` poisoned to zero on failure. Correct.
3. **API deletion mechanically complete** — files removed, header updated.

Previous non-blockers (dead includes, redundant `EVP_MD_CTX_reset`, stray comments, inconsistent EVP checking) are all fixed.

4 inline nits/suggestions below — none are blockers. Approving.

**Comment by tesseract-ripple** [comment]
**✅ Resolved in 69d810e1 + 814f94df.**

(Re-posting — the original inline comment on `src/commitments.c:32` is hidden as "outdated" after the line changed.)

Original issue: duplicate `#include <openssl/crypto.h>`. Fixed in 69d810e1 by replacing with `<openssl/evp.h>`, which introduced a duplicate `<openssl/evp.h>`. The second duplicate removed in 814f94df.

**Review by akinyele-ripple — APPROVED** [review]
Good to merge.
